README
##################
npg_ranger project
##################

Introduction
============

This project is a contribution towards development of the streaming API and servers under the
GA4GH http://ga4gh.org initiative (GA4GH Directory API and Streaming). The project has
both a server and a compatible http client.

The server conforms to as yet unpublished at the time of writing GA4GH API. It uses
the http transfer encoding protocol to stream sequencing data.

Package structure
=================

bin - node.js scripts, server and client
lib - node.js libraries
test - tests
docs - technical manuals
docs/api - generated API (run "grunt doc" to generate)
package.json - npm package definition
Gruntfile.js - grunt runner specification



CHANGES
LIST OF CHANGES for npg_ranger project
  - update CI to use only ubuntu latest
  - upgrade mongodb from 2.2.36 to 4.3.1
  - upgrade commander from 2.20.0 to 8.3.0
  - upgrade @prantlf/jsonlint from 8.0.2 to 10.2.0
  - upgrade async from 3.1.0 to 3.2.3
  - upgrade config-chain from 1.1.12 to 1.1.13
  - upgrade fs-extra from 8.1.0 to 10.0.0
  - upgrade http-shutdown from 1.2.1 to 1.2.2
  - upgrade moment from 2.24.0 to 2.29.1
  - upgrade semver from 6.3.0 to 7.3.5
  - upgrade send from 0.17.1 to 0.17.2
  - upgrade browserify from 16.5.0 to 17.0.0
  - upgrade decache from 4.5.1 to 4.6.1
  - upgrade grunt from 1.0.4 to 1.4.1
  - upgrade grunt-browserify from 5.3.0 to 6.0.0
  - upgrade grunt-jsdoc from 2.4.0 to 2.4.1
  - upgrade grunt-jsonlint from 2.0.0 to 2.1.3
  - upgrade pem from 1.14.3 to 1.14.4
  - upgrade tmp from 0.1.0 to 0.2.1
  - add GitHub Action and upgraded to minconda 4.10.3
  - add support for iRODS 4.2 clients in containers

release 3.0.0
  - upgrade node version to support node 10 and 12
    - remove support for node 6 and 8
    - update tests code to be compatible with new node versions
  - upgrade package dependencies and developer dependencies
  - modify test checking start up failure to improve test stability
  - added support for HTTP POST requests in server using header
    - encodes regions and passes it through headers
    - updated tests from GET that apply to POST
    - added new tests unique to POST
  - add a JSON schema that POST requests query must follow
  - implement new module to differentiate between valid multiregion
    queries and invalid duplicate values in queries
    - checks array of objects (valid multiregions) versus array of values (invalid)
    - created test spec to check functionality
  - implement new module to encode and decode the multiregions headers
    - gzips header and then converts to base64
    - created test spec to check functionality
  - implement new module to validate POST request query
    - checks if valid JSON first, then if fits the
      aforementioned JSON schema
    - checks range starts and ends
    - collapses any overlapping regions
    - created test spec to test functionality
  - update Docker for openstack
    - update ubuntu image to 18.04 from 14.04
    - update node version to 10.16.3 from 6.11.4
    - update mongo to 3.6.14 from 3.2.4
    - update apache httpd image to 2.4.41 from 2.4.27
    - update docker-compose version to 3 from 2
    - deploy bioinformatic tools using Conda
    - switch to newest versions of samtools,
      htslib, freebayes, biobambam
    - improve documentation
  - change installation method for dependencies in travis to use conda
    - add check for biobambam to also use cache
  - replace biobambam's "bamstreamingmarkduplicates" with samtools' "markdup"
    - still present in dev tests
  - fix existing bug in ranger client for outputting and piping into file
    - used to cause samtools errors in file and cause loggers to not display

release 2.0.0
  - upgrade async to 2.4.1 from 2.1.4
  - upgrade browserify to 14.4.0 from 13.3.0
  - upgrade fs-extra to 3.0.1 from 2.0.0
  - upgrade grunt-contrib-clean to 1.1.0 from 1.0.0
  - upgrade grunt-jasmine-nodejs to 1.6.0 from 1.5.4
  - upgrade http-shutdown to 1.2.0 from 1.1.0
  - upgrade moment to 2.18.1 from 2.17.1
  - upgrade mongodb to 2.2.29 from 2.2.22
  - upgrade pem to 1.9.7 from 1.9.4
  - upgrade request to 2.81.0 from 2.79.0
  - upgrade send to 0.15.3 from 0.14.2
  - update json response for server error reporting
  - update http error codes reported according to spec
    - httperror constructor does not take verbose anymore
    - trying to use unknown codes will result in 500 and log
    - remove 406
    - replace 409 & 422 with 400 + tailored status message
  - move trailer logic from error to controller
  - simplify url path to /ga4gh/sample (remove 'v.0.1/get')
  - upgrade samtools/htslib to 1.5
  - fix header key name for token authorisation
    - client sends token under 'Authorization' header
    - client formats token string as 'Bearer XXX'
    - server expects token under 'authorization' header (lowercase)
    - server tries to parse token string from 'Bearer XXX'
  - drop official support for node 4.x
  - update content-type to match spec
  - client warns on unsupported versions of json from server
  - set htsget as root element in JSON messages
  - update ranger Dockerfile to upgrade ranger docker to
    - httpd@2.4.27
    - npg_ranger@2.0.0
  - mark support for version ~1.0.0 of spec
  - fix complementary bugs in token use
    - server returns auth header for redirection
    - client only uses token from configuration for first request
  - remove legacy prefix/sufix code from htsget json proc
  - DataAccess sends access groups as array per file
  - don't try to suggest authurl when token missing because
    it may not make sense in complex configs
  - test with node 6.11.4 in travis matrix
  - support empty mediatype in data uri
  - switch to npm@4.5.0 in travis
  - set node@6.11.4 and npm@4.5.0 in package.json

release 1.4.2
  - fix not showing errors when next process in pipe closes
  - fix bug premature end of output stream when piping to output and end of
    stream in input (response)

release 1.4.1
  - update database entries in docker for public data
  - update ranger Dockerfile to upgrade ranger docker to
      - htslib@1.4.1,
      - samtools@1.4.1
      - npg_ranger@1.4.0

release 1.4.0
  - add options to enable https communication between npg_ranger and npg_sentry

release 1.3.0
  - parameter in CLI for CA
  - parameter in CLI for token configuration
  - use request library to support proxy env variables for CLI client
  - switch to streaming for CLI client
  - upgrade mongodb to 2.2.22 from 2.2.21
  - upgrade js-md5 to 0.4.2 from 0.4.1
  - upgrade winston to 2.3.1 from 2.3.0
  - upgrade send to 0.14.2 from 0.14.1
  - add new authorisation strategy; make authorisation requests to
    npg_sentry via tokens


lib/controller.js
"use strict";

/**
 * controller module.
 * @module lib/controller
 *
 * @description Controller module for the streaming HTTP server.
 *
 * @example <caption>Example usage of the controller.</caption>
 *   const RangerController = require('../lib/controller.js');
 *   // Create a new object, no authorisation.
 *   let c = new RangerController(request, response, db);
 *   // Create a new object with authorisation.
 *   let c1 = new RangerController(request, response, db, dir, true);
 *   // handle request
 *   c.handleRequest(host);
 *
 * @author Marina Gourtovaia
 * @copyright Genome Research Limited 2016
 */

const assert      = require('assert');
const fs          = require('fs');
const http        = require('http');
const os          = require('os');
const url         = require('url');
const LOGGER      = require('winston');
const DataAccess  = require('./auth.js');
const DataMapper  = require('./mapper.js');
const RangerModel = require('./model.js');
const HttpError   = require('./http/error.js');
const trailer     = require('./http/trailer.js');

const PROTOCOL_SEPARATOR = ':';
const DEFAULT_PROTOCOL   = 'http' + PROTOCOL_SEPARATOR;
const CUSTOM_FORWARDED_HEADER_NAME   = 'x-forwarded-host-suffix';
const FORWARDED_PROTOCOL_HEADER_NAME = 'x-forwarded-proto';

/** Application controller */
class RangerController {

  /**
   * Creates a RangerController object instance.
   * @param request  - HTTP request object, required
   * @param response - HTTP response object, required
   * @param db       - Mongo db handle, required
   * @param tmpDir   - directory path for temporary data, optional,
   *                   default to an OS tmp directory
   * @param skipAuth - boolen flag, optional, if true instructs to
   *                   skip authorisation
   */
  constructor(request, response, db, tmpDir, skipAuth) {

    var validate = (obj) => {
      return (obj && (typeof obj == 'object'));
    };

    assert(validate(request), 'HTTP request object is required');
    this.request = request;

    assert(validate(response), 'Server response object is required');
    assert((response instanceof http.ServerResponse),
      'Server response object is required');
    this.response = response;

    assert(validate(db), 'DB handle object is required');
    this.db = db;

    this.tmpDir = (tmpDir || os.tmpdir());
    assert(fs.existsSync(this.tmpDir),
      `Temp data directory '${this.tmpDir}' does not exist`);
    this.skipAuth = !!skipAuth;
  }

  contentType(format) {
    let primaryContentType = RangerModel.isTextualFormat(format) ? 'text' : 'application';
    return [primaryContentType, 'vnd.ga4gh.' + format.toLowerCase()].join('/');
  }

  errorResponse(response, code, m) {
    // Create eror object anyway, it will validate the input
    let err = new HttpError(response, code, m, true);
    if (!response.finished) {
      err.setErrorResponse();
      response.end();
    } else {
      throw new Error('Attempt to set error response on a closed stream');
    }
  }

  _setupPipeline(query) {
    assert(query);

    trailer.declare(this.response);
    this.response.setHeader("Content-Type", this.contentType(query.format));

    var endResponse = (truncated) => {
      trailer.setDataTruncation(this.response, truncated);
      this.response.end();
    };

    try {
      let model = new RangerModel(this.tmpDir);
      model.process(query, this.response, endResponse);
    } catch (ex) {
      if (!ex.toString().startsWith('Inconsistent format')) {
        throw ex;
      }
      this.errorResponse(this.response, 500, ex);
    }
  }

  _getData(user, host, query) {
    assert((user && host && query));
    var dm = new DataMapper(this.db);
    dm.once('error', (err) => {
      dm.removeAllListeners();
      this.errorResponse(this.response, 500, err);
    });
    dm.once('nodata', (message) => {
      dm.removeAllListeners();
      this.errorResponse(this.response, 404, message);
    });
    dm.once('data', (data) => {
      query.files = data.map( (d) => {return d.file;} );
      dm.removeAllListeners();
      if (this.skipAuth) {
        this._setupPipeline(query);
      } else {
        var da = new DataAccess(this.db);
        da.once('authorised', (username) => {
          da.removeAllListeners();
          LOGGER.debug(`User ${username} is given access`);
          this._setupPipeline(query);
        });
        da.once('failed', (username, message) => {
          da.removeAllListeners();
          this.errorResponse(this.response, 403,
            `Authorisation failed for user '${username}': ${message}`);
        });
        da.authorise(user.username, data.map( (d) => {return d.accessGroup;} ));
      }
    });
    dm.getFileInfo(query, host);
  }

  _checkFormat(format, requestFormat) {
    if ( !( format && requestFormat ) ) {
      throw new RangeError(`format and requestFormat must be defined`);
    }
    if (!RangerModel.supportsFormat(format)) {
      this.errorResponse(this.response, 409,
        `Format '${requestFormat}' is not supported, supported formats: ` +
        RangerModel.supportedFormats().join(', '));
      return false;
    }
    return true;
  }

  _ga4ghToDirectUrl(id, query) {
    LOGGER.debug('Process direct');
    LOGGER.debug(query);
    var _getInt = (str) => {
      /* Suppress unhelpful advice not to use Number as a constructor */
      /* jshint -W053 */
      let n = new Number(str);
      /* jshint +W053 */
      n = n.valueOf();
      if (!Number.isInteger(n)) {
        throw new RangeError(`'${str}' is not an integer`);
      }
      if (n < 0) {
        throw new RangeError(`'${str}' is not an unsigned integer`);
      }
      return n;
    };

    let url = `/sample?accession=${id}&format=${query.format}`;

    if (query.referenceName) {
      let refSeparator = encodeURIComponent(':');
      let region = query.referenceName;
      let start = query.start || '0';
      start = _getInt(start) + 1;
      if (query.end) {
        let end = _getInt(query.end) + 1;
        if (end < start) {
          throw new RangeError('Range end should be bigger that start');
        }
        region += (refSeparator + start + '-' + end);
      } else {
        if (start !== 1) {
          region += refSeparator + start;
        }
      }
      url += ('&region=' + region);
    } else {
      if ((typeof query.start !== 'undefined') || (typeof query.end !== 'undefined')) {
        throw new Error(
          "'referenceName' attribute requered if 'start' or 'end' attribute is given");
      }
    }

    return url;
  }

  _serverUrlFromRequest() {
    let urlObj = url.parse(this.request.url, false);
    ['query', 'search', 'path', 'pathname', 'href'].forEach((key) => {
      urlObj[key] = undefined;
    });
    urlObj.slashes  = true;
    urlObj.host     = this.request.headers.host;
    urlObj.protocol = DEFAULT_PROTOCOL;
    return url.format(urlObj);
  }

  _serverUrlFromForwardedHeaders() {
    let serverUrl;
    let key = 'x-forwarded-host';
    if (this.request.headers[key]) {
      serverUrl = this.request.headers[key];
      key = CUSTOM_FORWARDED_HEADER_NAME;
      if (this.request.headers[key]) {
        serverUrl += this.request.headers[key];
      }
      let protocol = this.request.headers[FORWARDED_PROTOCOL_HEADER_NAME]
        || DEFAULT_PROTOCOL;
      if (!protocol.endsWith(PROTOCOL_SEPARATOR)) {
        protocol += PROTOCOL_SEPARATOR;
      }
      serverUrl = protocol + '//' + serverUrl;
    }
    return serverUrl;
  }

  _redirect(id, query) {
    let url = '';
    LOGGER.debug('Process redirect');
    LOGGER.debug(query);
    try {
      url = this._ga4ghToDirectUrl(id, query);
    } catch (err) {
      this.errorResponse(this.response, 422, err.message);
    }
    if (url) {
      let serverUrl = this._serverUrlFromForwardedHeaders()
        || this._serverUrlFromRequest();
      let reply = {};
      reply.urls   = [serverUrl + url];
      reply.format = query.format;
      let redirect = JSON.stringify(reply);
      this.response.removeHeader('Transfer-Encoding');
      this.response.writeHead(200, 'OK, see redirection instructions in the body of the message', {
        'Content-Length': redirect.length,
        'Content-Type':   'application/json'});
      this.response.write(redirect);
      this.response.end();
    }
  }

  /**
   * Interprets a query url and its parameters.
   * Maps the query to data sources and authorises the client for access.
   * Handles request and sends a reply to the client.
   *
   * User authentication should be performed elsewhere. It is expected that the
   * 'x-remote-user' header is set in the request.
   *
   * User authorisation is per data resource, see 'auth' module for details.
   *
   * @param host - the name of the host to retrieve the data from.
   *               If the data are not available on the given host, the
   *               default host will be used.
   *
   * The following URLs are accepted:
   *   URL           Paramaters
   *   /file         name      - required
   *                 directory - optional
   *                 format    - required data format; 'SAM', 'CRAM' and 'BAM'
   *                             are recognised; if the format is not one of
   *                             these three, error response is returned; if the
   *                             format is not given, 'BAM' is used
   *                 region    - optional, in the format 'chr1:3000-3020', multiple
   *                             key-value pairs are allowed
   *   /sample       accession - sample accession number, required
   *                 region    - as for /file, but a single key-value pair only
   *                 format    - as for /file
   *   /ga4gh/v.0.1/get/sample/XXXX, where XXXX is a sample accession number (
   *                             the same as accession parameter in the /sample url)
   *                 format    - as above
   *                 referenceName - chromosome name, optional
   *                 start     - The start position of the range on the reference,
   *                             0-based inclusive; optional, if specified, referenceName
   *                             must also be specified
   *                 end       - The end position of the range on the reference,
   *                             0-based exclusive; optional, if specified, referenceName
   *                             must also be specified
   *
   * The URL starting with /ga4gh/v.0.1 outputs redirection JSON response according to
   * to the version of the GA4GH specification given in the URL. The rest of URLs stream the
   * data directly.
   *
   * Redirection JSON response contains a single absolute URL. In production this server is
   * likely to be running behind a reverse proxy. In this situation the JSON redirection respose
   * will contain the URL on the proxy server.
   *
   * The controller uses X-Forwarded-Host request header to determine the URL of the proxy server.
   * The protocol is inferred from X-Forwarded-Proto request header and defaults to http. If the
   * X-Forwarded-Host-Suffix request header is available, its value will be appended to the
   * server url (no additional separators are used).
   */
  handleRequest(host) {
    assert(host, 'The data host name is required');

    let user = {};
    user.username   = this.request.headers['x-remote-user'] || null;
    if (!user.username && !this.skipAuth) {
      return this.errorResponse(this.response, 401, 'Proxy authentication required');
    }

    let urlObj = url.parse(this.request.url, true);
    let path = urlObj.pathname;
    path = path ? path : '';
    let q = urlObj.query;

    let requestFormat;
    if (!q.format) {
      q.format = requestFormat = RangerModel.defaultFormat();
    } else {
      requestFormat = q.format;
      q.format = q.format.toUpperCase();
    }
    if (!this._checkFormat(q.format, requestFormat)) {
      return;
    }

    let directClback = (p) =>  {
      if (p === '/file' && !q.name) {
        this.errorResponse(this.response, 422,
          'Invalid request: file name should be given');
      } else if (p === '/sample' && !q.accession) {
        this.errorResponse(this.response, 422,
          'Invalid request: sample accession number should be given');
      } else {
        this._getData(user, host, q);
      }
    };

    let redirectClback = (p, id) =>  {
      if (!id) {
        this.errorResponse(this.response, 422,
          `Invalid request ${p}: sample accession number should be given`);
      } else {
        this._redirect(id, q);
      }
    };

    var matchPath = (p) => {
      let map = new Map();
      map.set(new RegExp(/^\/file$/),   directClback);
      map.set(new RegExp(/^\/sample$/), directClback);
      map.set(new RegExp(/^\/ga4gh\/v\.0\.1\/get\/sample\/([\.\w]+)$/), redirectClback);

      let it = map.entries();
      let el = it.next();
      let reResult = null;
      let callback = null;
      while (!reResult && el && el.done === false) {
        let re = el.value[0];
        reResult = re.exec(p);
        callback = el.value[1];
        el = it.next();
      }
      return !reResult ? false : {path: reResult[0], captured: reResult[1], cback: callback};
    };

    let match = matchPath(path);
    if (!match) {
      this.errorResponse(this.response, 404, 'URL not found : ' + path);
    } else {
      let cback = match.cback;
      cback(match.path, match.captured);
    }
  }
}

module.exports = RangerController;

